tree-sitter-stack-graphs/src/cli/match.rs

```diff
@@ -0,0 +1,161 @@
+            None => return Err(anyhow!("No stack graph language found")),
+        };
+        let source = file_reader.get(&self.source_path)?;
+        let tree = parse(lc.language, &self.source_path, source)?;
+        if self.stanza.is_empty() {
+            lc.sgl.tsg.try_visit_matches(&tree, source, true, |mat| {
+                print_matches(lc.sgl.tsg_path(), &self.source_path, source, mat)
+            })?;
+        } else {
+            for line in &self.stanza {
+                let stanza = lc
+                    .sgl
+                    .tsg
+                    .stanzas
+                    .iter()
+                    .find(|s| s.range.start.row <= line - 1 && line - 1 <= s.range.end.row)
+                    .ok_or_else(|| {
+                        anyhow!("No stanza on {}:{}", lc.sgl.tsg_path().display(), line)
+                    })?;
+                stanza.try_visit_matches(&tree, source, |mat| {
+                    print_matches(lc.sgl.tsg_path(), &self.source_path, source, mat)
+                })?;
+            }
+        }
+        Ok(())
+    }
+}
+
+fn print_matches(
+    tsg_path: &Path,
+    source_path: &Path,
+    source: &str,
+    mat: tree_sitter_graph::Match,
+) -> anyhow::Result<()> {
+    println!(
+        "{}: stanza query",
+        format!(
+            "{}:{}:{}",
+            tsg_path.display(),
+            mat.query_location().row + 1,
+            mat.query_location().column + 1
+        )
+        .bold(),
+    );
+    {
+        let full_capture = mat.full_capture();
+        print!("  matched ");
+        print_node(full_capture, true);
+        print_node_text(full_capture, source_path, source)?;
+        println!();
+    }
+    let width = mat
+        .capture_names()
+        .map(|n| n.len())
+        .max()
+        .unwrap_or_default();
+    if width == 0 {
+        return Ok(());
+    }
+    println!("  and captured");
+    for (name, quantifier, nodes) in mat.named_captures() {
+        let mut first = true;
+        for node in nodes {
+            if first {
+                first = false;
+                print!(
+                    "    @{}{}{} = ",
+                    name,
+                    quantifier_ch(quantifier),
+                    " ".repeat(width - name.len())
+                );
+            } else {
+                print!("     {}  | ", " ".repeat(width));
+            }
+            print_node(node, true);
+            print_node_text(node, source_path, source)?;
+            println!();
+        }
+    }
+    Ok(())
+}
+
+fn print_node_text(node: Node, source_path: &Path, source: &str) -> anyhow::Result<()> {
+    print!(", text: \"");
+    let text = node.utf8_text(source.as_bytes())?;
+    let summary: String = text
+        .chars()
+        .take(MAX_TEXT_LENGTH)
+        .take_while(|c| *c != '\n')
+        .collect();
+    print!("{}", summary.blue());
+    if summary.len() < text.len() {
+        print!("{}", "...".dimmed());
```

Comment:
Good move on indicating truncation with different font properties 👍🏻

Reply:
Upon further thought, should we use the Unicode ellipsis `…` here?

Reply:
Yeah, that's nice. It will make it even more clearly different from the code itself (unless it's Agda code full of Unicode 😬)

Reply:
One wrinkle is that in some (most?) fonts Unicode ellipsis renders as a single-width glyph e.g. `…` vs `...`

Reply:
@dcreager Do you think that's a problem?

Reply:
Oh just that I think it looks weird :smile:   That's not a valid reason to not use the Unicode character though.